ice-adapter/src/main/java/com/faforever/iceadapter/ice/PeerIceModule.java

```diff
@@ -95,8 +95,23 @@ private void createAgent() {
      */
     private void gatherCandidates() {
         log.info(getLogPrefix() + "Gathering ice candidates");
-        GameSession.getIceServers().stream().flatMap(s -> s.getStunAddresses().stream()).map(StunCandidateHarvester::new).forEach(agent::addCandidateHarvester);
-        GameSession.getIceServers().forEach(iceServer ->
+        List<IceServer> iceServers = GameSession.getIceServers();
+        OptionalDouble minRtt = iceServers.stream()
+                .map(server -> server.getRtt().join())
+                .filter(OptionalDouble::isPresent)
+                .mapToDouble(OptionalDouble::getAsDouble)
+                .min();
```

Comment:
Instead of getting the lowest RTT, you could map to new Map.Entry<IceServer, Double>() and then apply the following operations to the value (double) of the entry, for the min you can use a comparator. This way there's no need to search through the list again to find the element for the smallest rtt.

Reply:
Well the reason I'm searching through it again is in case there were any errors. If `ping` fails it doesn't necessarily mean that that candidate will fail, so I want to treat failures as "unknown" values and pass them through to the harvesting stage.